client/src/main/kotlin/io/spine/examples/chatspn/desktop/Client.kt

```diff
@@ -0,0 +1,414 @@
+            .select(MessageView::class.java)
+            .where(chat.queryFilter())
+            .orderBy(EntityColumn(whenPostedField), OrderBy.Direction.ASCENDING)
+            .run()
+        return messages
+    }
+
+    /**
+     * Subscribes actions on messages in the chat.
+     *
+     * @param chat id of the chat to subscribe on messages in
+     * @param updateAction an action that will be triggered when a new chat message is posted,
+     *                     or when an existing message is updated
+     * @param deleteAction an action that will be triggered when a message is deleted
+     */
+    public fun subscribeOnMessages(
+        chat: ChatId,
+        updateAction: (message: MessageView) -> Unit,
+        deleteAction: (messageDeleted: MessageMarkedAsDeleted) -> Unit
+    ) {
+        val updateSubscription = clientRequest()
+            .subscribeTo(MessageView::class.java)
+            .where(chat.stateFilter())
+            .observe(updateAction)
+            .post()
+        val deletionSubscription = clientRequest()
+            .subscribeToEvent(MessageMarkedAsDeleted::class.java)
+            .where(chat.eventFilter())
+            .observe(deleteAction)
+            .post()
+        messagesSubscriptions.add(updateSubscription)
+        messagesSubscriptions.add(deletionSubscription)
+    }
+
+    /**
+     * Cancel all subscriptions on messages in the chat.
+     */
+    public fun cancelMessagesSubscriptions() {
+        messagesSubscriptions.forEach { subscription ->
+            client.subscriptions().cancel(subscription)
+        }
+        messagesSubscriptions.clear()
+    }
+
+    /**
+     * Provides `ClientRequest` on behalf of `authenticatedUser` if it exists,
+     * or as guest if it doesn't.
+     */
+    private fun clientRequest(): ClientRequest {
+        if (null == authenticatedUser) {
+            return client.asGuest()
+        }
+        return client.onBehalfOf(authenticatedUser!!.id)
+    }
+
+    /**
+     * Subscribes an `action` to the provided event with ID.
+     */
+    public fun <E : EventMessage> subscribeToEvent(
+        id: Message,
+        event: Class<E>,
+        action: (event: E) -> Unit
+    ): Subscription {
+        val subscription = clientRequest()
+            .subscribeToEvent(event)
+            .where(EventFilter.eq(EventMessageField(Field.named("id")), id))
+            .observe(action)
+            .post()
+        return subscription
+    }
+
+    /**
+     * Subscribes an `action` to the provided event.
+     */
+    public fun <E : EventMessage> subscribeToEvent(
+        event: Class<E>,
+        action: (event: E) -> Unit
+    ): Subscription {
+        val subscription = clientRequest()
+            .subscribeToEvent(event)
+            .observe(action)
+            .post()
+        return subscription
+    }
+
+    /**
+     * Cancel the provided subscription.
+     */
+    public fun cancelSubscription(subscription: Subscription) {
```

Comment:
Current API is a bit patchy. We want both to shield users from dealing with "low-level" abstractions, and still want to provide more detail when dealing with events. 

Please rethink event subscriptions. Maybe, return an app-specific object, which would allow to call `cancel` on itself?